### PR TITLE
Add connection pooling for HTTP connections

### PR DESCRIPTION
**Pull request by abhin4v**
- connection pool is kept in the ConnectionData data.
- initConnectionData(Secure) take ConnectionPoolConfig as a parameter now.
- withConnection function now ges the connection from the pool instead of creating a new connection each time
- sets AutoClose to `No`
- added mmCloseSession function which closes the connection pool

Fixes issue #43.

### CODE REVIEW

File: src/Network/Mattermost/Types.hs [hunk 1]
```diff
@@ -49,38 +53,56 @@ maybeFail :: Parser a -> Parser (Maybe a)
...
 
+stripesCount :: Int
+stripesCount = 1
+
+idleConnTimeout :: NominalDiffTime
+idleConnTimeout = 30
+
+maxConnCount :: Int
+maxConnCount = 5
+
```
Comment (jtdaugherty) [inline src/Network/Mattermost/Types.hs]: Can you please make a new record data type containing these settings, such as `data PoolSettings`, provide a `defaultPoolSettings` with these values, and then make that a parameter to `initConnectionData(Insecure)`?

File: mattermost-api.cabal [hunk 2]
```diff
@@ -42,6 +42,7 @@ library
                      , aeson >= 1.0.0.0
                      , connection
                      , memory <0.14.3
+                     , resource-pool
```
Comment (jtdaugherty) [inline mattermost-api.cabal]: Can you please add a lower bound for this dependency? (The latest released major version is okay.)

File: src/Network/Mattermost/Types/Internal.hs [hunk 3]
```diff
@@ -33,11 +30,12 @@ autoCloseToHeader Yes = [mkHeader HdrConnection "Close"]
...
 data ConnectionData
   = ConnectionData
-  { cdHostname      :: Hostname
-  , cdPort          :: Port
-  , cdAutoClose     :: AutoClose
-  , cdConnectionCtx :: ConnectionContext
-  , cdToken         :: Maybe Token
-  , cdLogger        :: Maybe Logger
-  , cdUseTLS        :: Bool
+  { cdHostname       :: Hostname
```
Comment (jtdaugherty) [inline src/Network/Mattermost/Types/Internal.hs:33]: In the future, please make whitespace-only changes in separate commits.

File: src/Network/Mattermost/WebSocket.hs [hunk 4]
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE RecordWildCards #-}
```
Comment (jtdaugherty) [inline src/Network/Mattermost/WebSocket.hs]: Also, bear in mind that we tend not to use this extension much unless the benefit is significant. (Here it is minimal.)

**Comment by jtdaugherty** [comment]
Thanks for these changes! I've posted some comments. Also, @aisamanra is working on APIv4 changes to `mattermost-api` and will need to weigh in on whether this can be merged soon or need to wait.

**Comment by abhin4v** [comment]
@jtdaugherty I've made the changes you suggested.

**Comment by jtdaugherty** [comment]
Thanks!

**Pull request merged**